stdlib/importlib/readers.pyi

```diff
@@ -0,0 +1,60 @@
+# On py311+, things are actually defined in importlib.resources.readers,
+# and re-exported here,
+# but doing it this way leads to less code duplication for us
+
+import pathlib
+import sys
+from _typeshed import Incomplete, StrPath
+from collections.abc import Iterable, Iterator
+from io import BufferedReader
+from typing import NoReturn, TypeVar
+from typing_extensions import Literal, Never, Self
+
+if sys.version_info >= (3, 11):
+    import importlib.resources.abc as abc
+else:
+    import importlib.abc as abc
+
+if sys.version_info >= (3, 10):
+    if sys.version_info >= (3, 11):
+        __all__ = ["FileReader", "ZipReader", "MultiplexedPath", "NamespaceReader"]
+    else:
+        _T = TypeVar("_T")
+
+        def remove_duplicates(items: Iterable[_T]) -> Iterator[_T]: ...
+
+    class FileReader(abc.TraversableResources):
+        path: pathlib.Path
+        def __init__(self, loader) -> None: ...
+        def resource_path(self, resource: StrPath) -> str: ...
+        def files(self) -> pathlib.Path: ...
+
+    class ZipReader(abc.TraversableResources):
+        prefix: str
+        archive: Incomplete
+        def __init__(self, loader, module: str) -> None: ...
+        def open_resource(self, resource: str) -> BufferedReader: ...
+        def is_resource(self, path: StrPath) -> bool: ...
+        def files(self): ...
+
+    class MultiplexedPath(abc.Traversable):
+        def __init__(self, *paths: abc.Traversable) -> None: ...
+        def iterdir(self) -> Iterator[Self | abc.Traversable]: ...
+        def read_bytes(self) -> NoReturn: ...
+        def read_text(self, *args: Never, **kwargs: Never) -> NoReturn: ...  # type: ignore[override]
+        def is_dir(self) -> Literal[True]: ...
+        def is_file(self) -> Literal[False]: ...
+        if sys.version_info >= (3, 12):
+            def joinpath(self, *descendants: str) -> abc.Traversable: ...
+        else:
+            def joinpath(self, child: str) -> abc.Traversable: ...  # type: ignore[override]
+            __truediv__ = joinpath
+        def open(self, *args: Never, **kwargs: Never) -> NoReturn: ...  # type: ignore[override]
```

Comment:
```suggestion
        def open(self, *args: Unused, **kwargs: Unused) -> NoReturn: ...  # type: ignore[override]
```

`Never` would disallow arguably allowed calls like `.open("r")` (even if those always raise an exception, of course).

Reply:
Yes, that was a deliberate choice... I thought maybe it would be useful for type checkers to warn you if you provided arguments to this function. Since it always fails at runtime, it seems like you're probably making a mistake if you're passing arguments to it?

Reply:
But that should already be caught by the `NoReturn`.While this is a far-fetched example (using a non-existent typing feature), this disallows things like this:

```python
trav_open: SignatureOf[Traversible.open] = MultiplexedPath.open
```

I wonder why no type checker complains about using `Never`, since this breaks `Traversible`'s contract. 

That said, I don't feel strongly about this.

Reply:
> But that should already be caught by the `NoReturn`

Only if you have mypy's `--warn-unreachable` setting turned on (or the equivalent setting in other type checkers), and most people don't, unfortunately. Whereas annotating the arguments with `Never` means you'll get an error from the type checker even if you only have the default settings enabled.

I'll concede that this is a bit of a hack, and that we're really working around the lack of something like https://github.com/python/typing/issues/1043 -- but I think it's useful for the time being